### ISSUE TITLE
Msf.Server.Spawners.RegisterSpawnedProcess doesn't invoke callback

### ISSUE THREAD

**Issue by JadeWhite** [alvyxaz/barebones-masterserver#139]
I'm having issues getting a GameServer to register with a Master Server + Spawner on a cloud computer. I'm using a AmazonLightsail instance with a static IP, and all the ports on it are open (just for trying to solve this).

From the client, I contact the Spawner and request a new GameServer. However, right after the GameServer starts, in UnetServerStarter.cs, line 129:
      `Msf.Server.Spawners.RegisterSpawnedProcess(Msf.Args.SpawnId, Msf.Args.SpawnCode, (controller, error) =>...)`
       doesn't invoke the callback with a pass or fail result.

...and after 10 or so seconds, I get:
```
9/14/2017 6:52:04 PM|Fatal|WebSocket.<startReceiving>b__170_2:0|WebSocketSharp.WebSocketException: The header of a frame cannot be read from the stream.
                             at WebSocketSharp.WebSocketFrame.processHeader (System.Byte[] header) [0x00000] in <filename unknown>:0
                             at WebSocketSharp.WebSocketFrame+<>c__DisplayClass73_0.<readHeaderAsync>b__0 (System.Byte[] bytes) [0x00000] in <filename unknown>:0
                             at WebSocketSharp.Ext+<>c__DisplayClass54_0.<ReadBytesAsync>b__0 (IAsyncResult ar) [0x00000] in <filename unknown>:0
```
But I'm under the impression that comes from the GameServer shutting down after a timeout of some sort. 

It's especially strange because when I run everything on my computer with 127.0.0.1, I have no issues. My first thought was that my server IP's were wrong, but I've tried lots of variations to no avail. 

Has anyone encountered this issue before?

**Comment by JoaoBorks** [comment]
I had those before.. Are you using the MSF from the asset store or are you getting the latest version from the repository?

**Comment by JadeWhite** [comment]
This is the latest from the Asset store. I'll give the repository a shot....

**Comment by JadeWhite** [comment]
Not much different unfortunately... I'm getting the same results with the latest commit a137a9b.

**Comment by JoaoBorks** [comment]
If you print the status of the spawner on the client instance, what does it show?
[https://github.com/alvyxaz/barebones-masterserver/wiki/Spawning-Game-Servers](Ref:)
`while (request.Status != SpawnStatus.Finalized)
        {
            // Skip a frame, if it's still not finalized
            yield return null;

            // If status has changed
            if (currentStatus != request.Status)
            {
                // This should give it away
                Debug.Log("Status changed to: " + request.Status);
                currentStatus = request.Status;
            }
        }`

If we know the status, we have better tips of how to solve it

**Comment by JadeWhite** [comment]
The status reaches SpawnStatus.WaitingForProcess. I counted 30 seconds before it changes to Killed

**Comment by JoaoBorks** [comment]
Then at least it does change to Killed, maybe we're looking here to find an issue related to:

- Spawned Game Server Registration to Master Server
- Client connection to Game Server

Are you using the RoomTerminator? If you are, the room will close if empty. You could try to disable the RoomTerminator to track down if the problem is the client connecting to the Game Server or the Game Server Registration. Are you using the example spawner and game server scripts? If not, could you share them with us?

**Comment by JadeWhite** [comment]
I'm personally suspecting it's the former. I've laid down log messages all around the GameServer's 
MsfSpawnersServer.RegisterSpawnedProcess(), and the Master Server's SpawnerModule.HandleRegisterSpawnedProcess().

Looking at the logs of both,  RegisterSpawnedProcess() invokes the line where it sends the message, but HandleRegisterSpawnedProcess() doesn't get called. I'm almost beginning to suspect Ubuntu 16 is blocking the websockets for some reason.

I'm more or less using the same scripts used in the examples. UnetGameRoom, UnetServerStarter etc. I've extended some methods of the Unet Scripts to add events, but I don't think they would influence whether this particular message is sent.

I think I'm going to try and set up the example first, and see if that works.

**Comment by leonardochaia** [comment]
Hey guys, just wanted to add that I've successfully hosted on Ubuntu Server in AWS EC2.
All ports were open in the security group just for testing. I did this a while ago before getting really into the framework and had no issues.
I did host the master and spawner in the same container, running the QuickSetup scenes.

**Comment by JoaoBorks** [comment]
I'm also hosting my game server with customized MSF components on AWS EC2 Amazon Linux with no issues...

**Comment by JadeWhite** [comment]
Just did the QuickSetup example, and it works fine. Whoops, sorry guys.
Just have to find where my scripts are screwing things up now...

**Comment by JadeWhite** [comment]
Oh boy... I've been on a unpleasant ride trying to get this to work for the past few days.

Long story short, my scripts were flawless! It turned out to be a combination of several obtuse problems.

- - I followed https://stackoverflow.com/questions/25657596/how-to-set-up-gui-on-amazon-ec2-ubuntu-server to add a UI to the server (hopefully to get some visual, which wasn't fruitful). The end result made it so every time I tried to load my first level on the GameServer, it crashed instantly without an error message (even in headless mode). In the end I wiped my server and started over.
- -My first GameServer scene was so sufficiently complex that it would crash immediately if not built using BuildOptions.HeadlessMode.
- - My level music was butchering the framerate to a ridiculous amount on the server. This slowed down the thread so much, it kept my spawn status at SpawnStatus.WaitingForProcess. I only figured this out after creating a new scene and adding back the Gameobjects of my original scene one by one.
- - I had the Msf Gameobject + The Server Setup Gameobjects as a child of my NetworkManager in a giant prefab. For whatever reason, this prevented it from advancing past UnetGameRoom's RegisterRoom(). This eventually kept my spawn status at SpawnStatus.ProcessRegistered. I think this might have to do with how UnetGameRoom has a NetworkIdentity on it, and Unity doesn't like having NetworkIdentites as non-root prefabs. Not sure to be honest. 

It's fine to have your Msf gameobject + Children gameobjects as a prefab. Just don't have them be the children of a NetworkManager gameobject.

My work isn't complete, but at least I can play a round of my game now! Gonna mark this as resolved.
